gslab_scons/provenance.py

```diff
@@ -0,0 +1,195 @@
+                    line = '%s|%s' % (line, checksum)
+                file_details.append(line)
+   
+    return total_size, num_files, last_mtime, file_details, dirs
+
+
+def write_heading(start_path, provenance_path):
+    '''
+    Write standard heading for provenance: what and for where it is. 
+    '''
+    out = '*** GSLab directory provenance ***\ndirectory: %s\n' % os.path.abspath(start_path) 
+    with open(provenance_path, 'wb') as f:
+        f.write(out)
+
+    return None
+
+
+def write_directory_info(provenance_path, total_size, num_files, last_mtime):
+    '''
+    Write directory-level information to provenance. 
+    '''
+    out = 'total bytes: %s\n' % total_size + \
+          'number of files: %s\n' % num_files + \
+          'most recent modification time: %s\n' % last_mtime
+    with open(provenance_path, 'ab') as f:
+        f.write('\n*** Directory information\n')
+        f.write(out)
+
+    return None
+
+
+def write_readme(readme_path, provenance_path):
+    '''
+    Writes readme to provenance.
+    '''
+    with open(readme_path, 'rU') as f:
+        out = '%s' % f.read()
+    with open(provenance_path, 'ab') as f:
+        f.write('\n*** README verbatim\npath: %s\n' % os.path.abspath(readme_path))
+        f.write(out)
+
+    return None
+
+
+def write_detailed_info(provenance_path, details):
+    '''
+    Writes file-level information to provenance.
+    '''
+    out = '%s\n' % '\n'.join(details)
+    with open(provenance_path, 'ab') as f:
+        f.write('\n*** File information\n')
+        f.write(out)
+
+    return None
+
+
+def write_ending(provenance_path):
+    with open(provenance_path, 'ab') as f:
+        f.write('*' * 80 + '\n\n')
+    
+    return None
+
+
+def append_sub_provenance(root_provenance = './provenance.log',
+                          external_provenance = [],
+                          excluded_dirs = [],
+                          find_for_me = False,
+                          start_path = '.',
+                          verbose = False):
+    
+    files = misc.make_list_if_string(external_provenance)
+    if files == [] or find_for_me == True:
+        pattern = 'provenance*.log'
+        files += misc.finder(start_path, pattern, excluded_dirs)
+    files = sorted(files)
+    
+    
+    with open(root_provenance, 'a') as root_f:
+        for provenance in files:
+            if verbose == True:
+                print os.path.abspath(provenance)
+            
+            warning_message = '\nThe file %s does not appear to be a GSLab provenance file.\n' % provenance
+            warn = False
+            
+            if os.path.normpath(root_provenance) != os.path.normpath(provenance):
+                root_f.write('*** Subdirectory provenance: %s\n' % os.path.abspath(provenance))
+                with open(provenance, 'rU') as f:
+                    
+                    try:
+                        content = f.readlines()
```

Comment:
Move readlines below file opening. Move try/except block to same indentation as `with`.